### PR TITLE
Use edge function to handle placeholder fallbacks

### LINKED ISSUE #4707: Placeholder redirects and CORS don't mix
## The problem

We want to have the service worker cache brand images for performance and offline use (https://github.com/home-assistant/frontend/pull/17840).  In order to do this, all  responses must contain the `Access-Control-Allow-Origin` header.  These were added in #4652, but they do not apply to the 302 redirects to the placeholder images.  So when the frontend goes to fetch them, a CORS error is thrown.

AFAIK, there is no way to actually modify the headers of a redirect for security reasons (e.g. [`Response.redirect()` creates a response with immutable headers](https://developer.mozilla.org/en-US/docs/Web/API/Fetch_API/Basic_concepts#guard)).  There is also a [relevant discussion in the Netlify support forum](https://answers.netlify.com/t/netlify-does-not-send-custom-headers-in-a-redirect-file-using-https-in-the-to/25773/15).

A potentially easy fix is to just rewrite the URL instead of redirecting it, which [can be done in `netlify.toml`](https://docs.netlify.com/routing/redirects/rewrites-proxies/).  However, Netlify applies the `[[headers]]` for the request URL, not the rewritten one, so it won't be cached as desired.

Similarly, `[[headers]]` in the config file are applied regardless of the response status, which means requests without a placeholder fallback that return a 404 are currently being cached for the longer time.

The only way I see to achieve the desired responses is to use  an edge function.

## Proposed Solution
1. Use an edge function to handle all the image responses.  It sets the cache and CORS headers as desired.
2. Use a query parameter to request a placeholder fallback instead of the `/_/` directory.  This cuts the domain storage in half, and makes it much simpler and safer to test for in the edge function.  It would also make it easier for the frontend service worker to manipulate how images are cached if needed.

### PR DESCRIPTION
**Pull request by steverep**
## Proposed change

Fixes #4707 by implementing proposed solution in the issue.  After testing, will follow up with PRs for frontend and docs to use new URL search parameter.
I used a deno devcontainer to develop the function.  I can push that config if desired.

## Type of change

- [ ] Add a new logo or icon for a new core integration
- [ ] Add a missing icon or logo for an existing core integration
- [ ] Add a new logo or icon for a custom integration (custom component)
  - [ ] I've opened up a PR for my custom integration on the [Home Assistant
    Python wheels repository](https://github.com/home-assistant/wheels-custom-integrations)
- [ ] Replace an existing icon or logo with a higher quality version
- [ ] Removing an icon or logo

## Additional information

- This PR fixes or closes issue: fixes #4707 
- Link to code base pull request: 
- Link to documentation pull request: 
- Link to integration documentation on our website: 

## Checklist

- [ ] The added/replaced image(s) are **PNG**
- [ ] Icon image size is 256x256px (`icon.png`)
- [ ] hDPI icon image size is 512x512px for  (`icon@2x.png`)
- [ ] Logo image size has min 128px, but max 256px, on the shortest side (`logo.png`)
- [ ] hDPI logo image size has min 256px, but max 512px, on the shortest side (`logo@2x.png`)

### CODE REVIEW

**Comment by steverep** [comment]
Everything works as expected in the deploy preview.  Tomorrow I will add rules so that `_/` URLs continue to work for existing HA versions.

File: README.md [hunk 1]
```diff
@@ -22,65 +22,51 @@ Each of these two main folders contain domain folders. Each domain folder is
...
-- `https://brands.home-assistant.io/[domain]/icon.png`
-- `https://brands.home-assistant.io/[domain]/logo.png`
-- `https://brands.home-assistant.io/[domain]/icon@2x.png`
-- `https://brands.home-assistant.io/[domain]/logo@2x.png`
-- `https://brands.home-assistant.io/_/[domain]/icon.png`
-- `https://brands.home-assistant.io/_/[domain]/logo.png`
-- `https://brands.home-assistant.io/_/[domain]/icon@2x.png`
-- `https://brands.home-assistant.io/_/[domain]/logo@2x.png`
+```
+https://brands.home-assistant.io/[domain]/{dark_}[icon|logo]{@2x}.png
```
Comment (frenck) [inline README.md:35]: Please revert this. This might work for a programmer, but lots of users don't understand this at all.

This is also completely unrelated to the change provided.

File: netlify/edge-functions/placeholder-fallback.ts [hunk 2]
```diff
@@ -0,0 +1,38 @@
+import type { Context } from "https://edge.netlify.com";
```
Comment (frenck) [inline netlify/edge-functions/placeholder-fallback.ts:1]: We can't use edge functions. These will costs us a lot of money.

**Review by frenck — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (steverep) [reply README.md:35]: The relation was to be consistent when showing a URL with the search parameter below, without repeating a list of 8 nearly identical lines.  Could just do something like `.../[domain]/[image]` instead?

File: netlify/edge-functions/placeholder-fallback.ts (hunk 2 above)
Comment (steverep) [reply netlify/edge-functions/placeholder-fallback.ts:1]: Ugh... I just stupidly assumed the open source plan was unlimited invocations.  Apparently not even the enterprise plan is unlimited.

We could potentially keep the invocations under the  limit by caching the function responses (invocation only occurs when it runs per their docs).  Search parameter would then be out though as they only cache based on path name.  Theoretically, invocations would then = number of files x 2 x deploys per month.

Working within the config file constraints, the best alternative would be to switch the redirects to rewrites and lose the ability to vary the cache control header.

Comment (steverep) [reply netlify/edge-functions/placeholder-fallback.ts:1]: > This is a frontend problem, solve it in the frontend 

The frontend cannot control responses from https://brands.home-assistant.io.  A solution there means unnecessary extra code, repeating the functionality here, and, most notably, wasted network requests that reduce performance.

> (that said, I don't see a problem at all, so the "problem" in that sentence is overkill).

I did my best to explain it in the issue.  "Sorry to say", but you not seeing a problem doesn't mean it doesn't exist.  If you'd like to see it, pull the frontend PR, test it, and look at the errors in the browser console.

**Comment by steverep** [comment]
Given Netlify's invocation limits and billing on edge functions, closing in favor of #4722
